src/posix/platform/settings.cpp

```diff
@@ -32,429 +32,442 @@
+    {
+        uint16_t key;
+        uint16_t length;
+        ssize_t  rval;
 
-    sSettingsBaseAddress =
-        (swapAddress == SETTINGS_CONFIG_BASE_ADDRESS) ? (swapAddress + settingsSize) : SETTINGS_CONFIG_BASE_ADDRESS;
+        rval = read(sFd, &key, sizeof(key));
+        VerifyOrExit(rval == sizeof(key), error = OT_ERROR_PARSE);
 
-    initSettings(sSettingsBaseAddress, static_cast<uint32_t>(kSettingsInSwap));
-    sSettingsUsedSize = kSettingsFlagSize;
-    swapAddress += kSettingsFlagSize;
+        rval = read(sFd, &length, sizeof(length));
+        VerifyOrExit(rval == sizeof(length), error = OT_ERROR_PARSE);
+
+        offset += sizeof(key) + sizeof(length) + length;
+        VerifyOrExit(offset == lseek(sFd, length, SEEK_CUR), error = OT_ERROR_PARSE);
+    }
 
-    while (swapAddress < (oldBase + usedSize))
+exit:
+    if (error == OT_ERROR_PARSE)
     {
-        OT_TOOL_PACKED_BEGIN
-        struct addSettingsBlock
-        {
-            struct settingsBlock block;
-            uint8_t              data[kSettingsBlockDataSize];
-        } OT_TOOL_PACKED_END addBlock;
-        bool                 valid = true;
+        VerifyOrDie(ftruncate(sFd, 0) == 0);
+    }
+}
 
-        utilsFlashRead(swapAddress, reinterpret_cast<uint8_t *>(&addBlock.block), sizeof(struct settingsBlock));
-        swapAddress += sizeof(struct settingsBlock);
+otError otPlatSettingsGet(otInstance *aInstance, uint16_t aKey, int aIndex, uint8_t *aValue, uint16_t *aValueLength)
+{
+    otError     error  = OT_ERROR_NOT_FOUND;
+    const off_t size   = lseek(sFd, 0, SEEK_END);
+    off_t       offset = lseek(sFd, 0, SEEK_SET);
 
-        if (!(addBlock.block.flag & kBlockAddCompleteFlag) && (addBlock.block.flag & kBlockDeleteFlag))
-        {
-            uint32_t address = swapAddress + getAlignLength(addBlock.block.length);
+    OT_UNUSED_VARIABLE(aInstance);
+    VerifyOrExit(offset == 0 && size >= 0, error = OT_ERROR_PARSE);
 
-            while (address < (oldBase + usedSize))
-            {
-                struct settingsBlock block;
+    while (offset < size)
+    {
+        uint16_t key;
+        uint16_t length;
+        ssize_t  rval;
 
-                utilsFlashRead(address, reinterpret_cast<uint8_t *>(&block), sizeof(block));
+        rval = read(sFd, &key, sizeof(key));
+        VerifyOrExit(rval == sizeof(key), error = OT_ERROR_PARSE);
 
-                if (!(block.flag & kBlockAddCompleteFlag) && (block.flag & kBlockDeleteFlag) &&
-                    !(block.flag & kBlockIndex0Flag) && (block.key == addBlock.block.key))
+        rval = read(sFd, &length, sizeof(length));
+        VerifyOrExit(rval == sizeof(length), error = OT_ERROR_PARSE);
+
+        if (key == aKey)
+        {
+            if (aIndex == 0)
+            {
+                error = OT_ERROR_NONE;
+
+                if (aValueLength)
                 {
-                    valid = false;
-                    break;
+                    assert(*aValueLength >= length);
```

Comment:
I don't think we need this assert

`otPlatSettingsGet()` can be called with a `*aVlaueLength` being smaller than the value len.
In particular, it can be called with `*aValueLength = 0` and `aValue=NULL` to get the length.